site/content/blog/2024-05-21-introducing-the-grpc-addon/index.md

```diff
@@ -0,0 +1,100 @@
+
+This add-on aims to simplify the testing process by providing tools for:
+
+* **Encoding and Decoding** : Easily encode and decode Protobuf messages for testing purposes.
+* **Testing gRPC Endpoints** : You can combine it with existing ZAP functionality to abuse gRPC endpoints in many different ways.
+
+### Installation
+
+If you are using the latest version of ZAP then you can download the gRPC add-on from the 
+[ZAP Marketplace](/addons/).
+
+### How to Display a gRPC Decoded Message
+
+In the Message View panel, change the Body to gRPC.
+
+![Displaying a gRPC decoded message](images/gRPCResponseBody.png)
+
+### Understanding Decoded Protobuf Messages
+
+When decoding Protobuf messages, it's essential to understand the structure of the decoded data. 
+Protobuf messages consist of fields, each with a field number, wire type, and value. 
+Let's delve deeper into these components:
+
+#### Field Structure
+
+* **Field Number** : Each field in a Protobuf message is assigned a unique field number. This number is used to identify and distinguish between different fields within the message.
+* **Wire Type** : The wire type determines how the field's value is encoded and provides information about the type of data stored in the field.
+* **Value** : The actual data contained within the field, encoded according to its wire type.
+
+### Protobuf Message Structure example
+
+Let's understand the decoded Protobuf Message Structure using an example.
+
+This is the Base64 encoded form of the binary text received by ZAP.
+
+```
+AAAAAIYJZCZMzMzcXkAVrseHQhi5YCDqrcDlJCixqAM1QEIPADOQSGVsbG8sIFByb3RvYnVmIUJMCgs
+xMjMgTWFpbiBTdBIGTX1DaXR5GgUxMjMONSIuCgtIZWxsbyBXb3JsZBIITXkgV29ybGQaCllvdXIgV2
+9ybGQiCU91ciBXb3JsZEjqrcDLJA==
+```
+
+On the left side, we have the original message retrieved by the gRPC endpoint in the backend. On the right side, we have the decoded message from the above encoded text, processed by the ZAP gRPC addon.
+
+![Encoded an decoded gRPC message](images/decoded.png)
+
+* Each field in the message is represented as `field number:wire type::value`.
+* Each key-value pair starts on a new line.
+* Additional specifiers are only used for double (`1D`), float (`5F`), and nested messages (`2N`).
+* For wire type `2`, the value should be enclosed in double quotes, except for nested messages (`2N`).
+* For nested messages(`2N`):
+  *  After the field number and wire type (`fieldNumber:2N::`), an open curly brace is placed and start adding nested message fields from new line.
+  * The nested message is closed with a closing curly brace on a separate new line.
+* Please don't use unnecessay space.
+
+### Writing and Editing Decoded Protobuf Messages
+
+You can only edit this Decoded Message in [Requester](/docs/desktop/addons/requester/tab/) and [Manual Request Editor](/docs/desktop/addons/requester/dialogs/) windows - 
+simply change the part after (`::`) which contain value or you can write new fields by following above mentioned rules.
```

Comment:
...the above mentioned rules


(Add "the")

Also for that part (rules above): Does it matter if new lines are LF or CRLF? Should there be an example to support the curly braces explanation?

Reply:
Added ths 'the'. 
I'll let @amitpanwar789 answer about the explanation...

Reply:
In the algorithm I am using '\n' to extract the key value pair from the string of the message panel.

Reply:
It should be LF

Reply:
Do we need to clarify that or am I overthinking it? 😉

Reply:
The editor will use LF, maybe the help is a better place to go into details.

Reply:
If we are talking about gui then I guess the user will just hit enter

Reply:
so ok to leave as is?